posthog/storage/object_storage.py

```diff
@@ -74,6 +89,49 @@ def read_bytes(self, bucket: str, key: str) -> Optional[bytes]:
             logger.error("object_storage.read_failed", bucket=bucket, file_name=key, error=e, s3_response=s3_response)
             raise ObjectStorageError("read failed") from e
 
+    def list_all_objects(self, bucket: str, prefix: str) -> Optional[List[str]]:
+        objects = []
+        try:
+            has_next = True
+            while has_next:
+                last_key = objects[-1]["Key"] if len(objects) > 0 else ""
+                s3_response = self.aws_client.list_objects_v2(
+                    Bucket=bucket, Prefix=prefix, Delimiter="/", StartAfter=last_key
+                )
+                has_next = s3_response["IsTruncated"]
+                objects.extend(s3_response["Contents"])
+            return objects
+        except Exception as e:
+            logger.error("object_storage.list_all_objects_failed", bucket=bucket, prefix=prefix, error=e)
+            raise ObjectStorageError("list_all_objects failed") from e
+
+    def _read_for_threading(self, result: List, index: int, bucket: str, key: str):
+        result[index] = self.read(bucket, key)
+
+    def read_all(self, bucket: str, keys: List[str], max_concurrent_requests: int) -> Optional[List[Tuple[str, str]]]:
```

Comment:
Theoretically, this should be fast, but in practice it is the opposite. I'm not sure why

Reply:
I guess for the metadata we don't do any progressive loading so we can't make the time to first byte really fast and therefore start playing immediately?

I'm not sure I'll be able to properly look at this today as am flying. Seems worth digging in first to where the bottleneck here is.

If all else fails we may be able to take advantage of that there is only one consumer processing messages that we can compact metadata down to one file 🤔